clconf/template/template_funcs.go

```diff
@@ -281,3 +284,91 @@ func RegexReplace(regex, src, repl string) (string, error) {
 	}
 	return re.ReplaceAllString(src, repl), nil
 }
+
+// sortType accepts an array because the input is an optional name of the type
+// for sorting and the actual implementation methods (getsvs, getksvs) accept
+// varargs so this utility function allows direct call without unpacking.
+func sortType(input []string) (string, error) {
+	r := "string"
+	if len(input) > 0 {
+		r = input[0]
+	}
+	if r != "string" && r != "int" {
+		return "", fmt.Errorf("sort: Type '%s' is not supported (only int, string)", r)
+	}
+	return r, nil
+}
+
+type asInt []string
+
+func (p asInt) Len() int { return len(p) }
+func (p asInt) Less(i, j int) bool {
+	a, aerr := strconv.Atoi(p[i])
+	b, berr := strconv.Atoi(p[j])
+
+	if aerr == nil {
+		if berr == nil {
+			return a < b
+		}
+		return true // Numbers come first
+	} else {
+		if berr == nil {
+			return false
+		}
+	}
+	return p[i] < p[j]
+}
+func (p asInt) Swap(i, j int) { p[i], p[j] = p[j], p[i] }
+
+// sortCopy sorts a copy of the input as specified type (string, int, default: string).
+func sortCopy(v []string, asType ...string) ([]string, error) {
+	sortType, err := sortType(asType)
+	if err != nil {
+		return nil, err
+	}
+
+	sorted := make([]string, len(v))
+	copy(sorted, v)
+	if sortType == "int" {
+		sort.Sort(asInt(sorted))
+	} else {
+		sort.Strings(sorted)
+	}
+	return sorted, nil
+}
+
+func getsvs(s *memkv.Store) func(string, ...string) ([]string, error) {
+	return func(pattern string, asType ...string) ([]string, error) {
+		vals, err := s.GetAllValues(pattern)
+		if err != nil {
+			return nil, err
+		}
+		return sortCopy(vals, asType...)
+	}
+}
+
+func getksvs(s *memkv.Store) func(string, ...string) ([]string, error) {
+	return func(pattern string, asType ...string) ([]string, error) {
+		ks, err := s.GetAll(pattern)
+		if err != nil {
+			return nil, err
+		}
+		kvMap := make(map[string]string)
+		keys := make([]string, len(ks))
+		r := make([]string, len(ks))
+		for i, kv := range ks {
+			key := kv.Key[strings.LastIndex(kv.Key, "/")+1:]
+			kvMap[key] = kv.Value
+			keys[i] = key
+		}
+
+		keys, err = sortCopy(keys, asType...)
```

Comment:
This doesn't make any sense to me.  Why would you make a _copy_ only to assign it back to itself?  That seems like (possibly) a lot of work for no benefit.  The only reason I can think of to create a _copy_ is if you wanted to leave the original alone, but assigning back to the original (which is a local var) means you do not care about the original any longer, so why not sort in place?

Reply:
[Discussion here](https://github.com/pastdev/clconf/pull/26#discussion_r573847962)